embedded/store/immustore.go

```diff
@@ -2737,3 +2828,163 @@ func minUint64(a, b uint64) uint64 {
+
+	hdr, err := tdr.readHeader(s.maxTxEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	if hdr.NEntries < index {
+		return nil, ErrCorruptedTxDataMaxTxEntriesExceeded
+	}
+
+	e := &TxEntry{k: make([]byte, s.maxKeyLen)}
+
+	for i := 0; i < index; i++ {
+		err = tdr.readEntry(e)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return e, nil
+}
+
+// TruncateUptoTx deletes the value-log file up to transactions
+// that are strictly below the specified minTxID.
+func (s *ImmuStore) TruncateUptoTx(minTxID uint64) error {
+	/*
+		When values are appended to the value log file, they are not
+		inserted in strict monotic order of time. Depending on the
+		maxConcurrency value, there could be n transactions trying
+		to insert into the log at the same time. This could lead to
+		the situation where a transaction (n+1) could be inserted
+		in the value log before transaction (n)
+						  discard point
+								|
+								v
+				--------+-------+--------+----------
+						|       |        |
+					tn+1:vx   tn:vx  tn-1:vx
+						|                |
+						+----------------+
+						max concurrency
+							range
+		If the log is truncated upto tn, it could lead to removal of
+		data for tn+1. To avoid this overlap, we first go back from
+		the discard point to fetch offsets across vlogs for safe delete
+		points, and then check for transactions further than the
+		discard point to figure out the least offset from the range
+		which can safely be deleted. This offset, tn+1 in the above
+		scenario is the safest point to discard upto to avoid deletion
+		of values for any future transaction.
+	*/
+
+	s.logger.Infof("running truncation up to transaction '%d'", minTxID)
+
+	var err error
+	// tombstones maintain the minimum offset for each value log file that can be safely deleted.
+	tombstones := make(map[byte]int64)
+
+	readFirstEntryOffset := func(id uint64) (*TxEntry, error) {
+		return s.readTxOffsetAt(id, false, 1)
+	}
+
+	back := func(id uint64) error {
+		first, err := readFirstEntryOffset(id)
+		if err != nil {
+			return err
+		}
+
+		// Iterate over all past transactions and store the minimum offset for each value log file.
+		v, off := decodeOffset(first.VOff())
+		if _, ok := tombstones[v]; !ok {
+			tombstones[v] = off
+		}
+		return nil
+	}
+
+	front := func(id uint64) error {
+		first, err := readFirstEntryOffset(id)
+		if err != nil {
+			return err
+		}
+
+		// Check if any future transaction offset lies before past transaction(s)
+		// If so, then update the offset to the minimum offset for that value log file.
+		v, off := decodeOffset(first.VOff())
+		if val, ok := tombstones[v]; ok {
+			if off < val {
+				tombstones[v] = off
+			}
+		}
+		return nil
+	}
+
+	// Walk back to get offsets across vlogs which can be deleted safely.
+	// This way, we can calculate the minimum offset for each value log file.
+	{
+		var i uint64 = minTxID
+		for i > 0 && len(tombstones) != s.MaxIOConcurrency() {
+			err = back(i)
+			i--
```

Comment:
shouldn't i be decremented if there is no error?

Reply:
actually the error is being returned, so should not matter, but yeah can be moved post error check